File: python/src/runapi/suno/resources/add_vocals.py
"""Suno add_vocals resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource

from .. import _validators
from ..types import CompletedAddVocalsResponse, AddVocalsResponse


class AddVocals(Resource):
    """Add vocals to uploaded audio."""

    ENDPOINT = "/api/v1/suno/add_vocals"

    RESPONSE_CLASS = AddVocalsResponse
    COMPLETED_RESPONSE_CLASS = CompletedAddVocalsResponse

    def run(self, **params: Any) -> Any:
        """Generate and add vocals and poll until it completes.

        Args:
            **params: add-vocals parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an add-vocals task and return immediately with an id.

        Args:
            **params: add-vocals parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an add-vocals task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_add_vocals(params)

File: ruby/lib/runapi/suno/resources/add_vocals.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Generates and adds vocals to an uploaded instrumental track.
      class AddVocals
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/add_vocals"
        RESPONSE_CLASS = Types::AddVocalsResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedAddVocalsResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          Validators.validate_add_vocals!(params, self)
        end
      end
    end
  end
end
